From a465ebd93741b927b25e59b409e5042cc4a2a460 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:00:50 -0400
Subject: [ethernet] mlx4: Fix smatch error - possible access to a null variable

Message-id: <dc995182d6913b54511ac60776205dfbbc34202a.1410425016.git.avadai@redhat.com>
Patchwork-id: 93719
O-Subject: [Patch RHEL7.1 V1 114/197] net/mlx4_core: Fix smatch error - possible access to a null variable
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Eyal Perry <eyalpe@mellanox.com>

Fix the "error: we previously assumed 'out_param' could be null" found
by smatch semantic checker on:
drivers/net/ethernet/mellanox/mlx4/cmd.c:506 mlx4_cmd_poll()
drivers/net/ethernet/mellanox/mlx4/cmd.c:578 mlx4_cmd_wait()

Signed-off-by: Eyal Perry <eyalpe@mellanox.com>
Signed-off-by: Amir Vadai <amirv@mellanox.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit c05a116f397242e0ed4353bb4d36671a1f86ec77)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/cmd.c b/drivers/net/ethernet/mellanox/mlx4/cmd.c
index 2420103..357dcb0 100644
--- a/drivers/net/ethernet/mellanox/mlx4/cmd.c
+++ b/drivers/net/ethernet/mellanox/mlx4/cmd.c
@@ -473,6 +473,13 @@ static int mlx4_cmd_poll(struct mlx4_dev *dev, u64 in_param, u64 *out_param,
   goto out;
  }
 
+ if (out_is_imm && !out_param) {
+  mlx4_err(dev, "response expected while output mailbox is NULL for command 0x%x\n",
+    op);
+  err = -EINVAL;
+  goto out;
+ }
+
  err = mlx4_cmd_post(dev, in_param, out_param ? *out_param : 0,
        in_modifier, op_modifier, op, CMD_POLL_TOKEN, 0);
  if (err)
@@ -551,6 +558,13 @@ static int mlx4_cmd_wait(struct mlx4_dev *dev, u64 in_param, u64 *out_param,
  cmd->free_head = context->next;
  spin_unlock(&cmd->context_lock);
 
+ if (out_is_imm && !out_param) {
+  mlx4_err(dev, "response expected while output mailbox is NULL for command 0x%x\n",
+    op);
+  err = -EINVAL;
+  goto out;
+ }
+
  init_completion(&context->done);
 
  mlx4_cmd_post(dev, in_param, out_param ? *out_param : 0,
-- 
1.7.1